src/WebForms/UI/WebControls/Panel.cs

```diff
@@ -0,0 +1,361 @@
+            }
+            else {
+                writer.AddStyleAttribute(HtmlTextWriterStyle.WhiteSpace, "nowrap");
+            }
+        }
+
+        if (Direction == ContentDirection.LeftToRight) {
+            writer.AddAttribute(HtmlTextWriterAttribute.Dir, "ltr");
+        }
+        else if (Direction == ContentDirection.RightToLeft) {
+            writer.AddAttribute(HtmlTextWriterAttribute.Dir, "rtl");
+        }
+
+        if (!DesignMode &&
+            (Page != null) &&
+            (Page.RequestInternal != null)
+            //&&
+            //TODO https://github.com/twsouthwick/systemweb-adapters-ui/issues/30
+           // (Page.Request.Browser.EcmaScriptVersion.Major > 0) &&
+           // (Page.Request.Browser.W3CDomVersion.Major > 0)
+           )
+           {
+            if (DefaultButton.Length > 0) {
+                // Find control from the page if it's a hierarchical ID.
+                // Dev11 bug 19915
+                Control c = FindControlFromPageIfNecessary(DefaultButton);
+
+                if (c is IButtonControl) {
+                    Page.ClientScript.RegisterDefaultButtonScript(c, writer, true /* UseAddAttribute */);
+                }
+                else {
+                    throw new InvalidOperationException(SR.GetString(SR.HtmlForm_OnlyIButtonControlCanBeDefaultButton, ID));
+                }
+            }
+        }
+    }
+
+    /// <devdoc>
+    ///    [To be supplied.]
+    /// </devdoc>
+    private void AddScrollingAttribute(ScrollBars scrollBars, HtmlTextWriter writer) {
+        switch (scrollBars) {
+            case ScrollBars.Horizontal:
+                writer.AddStyleAttribute(HtmlTextWriterStyle.OverflowX, "scroll");
+                break;
+            case ScrollBars.Vertical:
+                writer.AddStyleAttribute(HtmlTextWriterStyle.OverflowY, "scroll");
+                break;
+            case ScrollBars.Both:
+                writer.AddStyleAttribute(HtmlTextWriterStyle.Overflow, "scroll");
+                break;
+            case ScrollBars.Auto:
+                writer.AddStyleAttribute(HtmlTextWriterStyle.Overflow, "auto");
+                break;
+            default:
+                break;
+        }
+    }
+
+
+    /// <internalonly/>
+    /// <devdoc>
+    ///    <para>A protected method. Creates a panel control style.</para>
+    /// </devdoc>
+    protected override Style CreateControlStyle() {
+        return new PanelStyle(ViewState);
+    }
+
+
+    /// <internalonly/>
+    public override void RenderBeginTag(HtmlTextWriter writer) {
+        AddAttributesToRender(writer);
+
+        HtmlTextWriterTag tagKey = TagKey;
+        if (tagKey != HtmlTextWriterTag.Unknown) {
+            writer.RenderBeginTag(tagKey);
+        }
+        else {
+            writer.RenderBeginTag(TagName);
+        }
+
+        string s = GroupingText;
+        bool useGrouping = (s.Length != 0) && !(writer is Html32TextWriter);
```

Comment:
FYI, aliasing Html32TextWriter to HtmlTextWriter will affect lines like this.

Reply:
yeah, that's why keeping it as an alias for now so we can easily change it back later if needed. Html32TextWriter (according to the docs) enabled things for Html 3.2 which is an old standard by now, so I'd rather not bring it in until we *need* it. HtmlTextWriter is at least 4.0 (still old by now), but what we should default to unless there's a compelling reason not to